### DISCUSSION TITLE
How to diagnose "SNAPSHOT can not be represented in Maven model and will not be visible to non-OSGi aware Maven plugins" warning?

### DISCUSSION THREAD

**Discussion by sewe** [eclipse-tycho/tycho#877] [Q&A]
Hi,

recently (after a bigger Tycho upgrade), I get the following warning at the beginning of my build during target platform resolution:

```
[WARNING] eclipse-plugin:com.google.guava:21.0.0.v20170206-1425: /home/sewe/.m2/repository/org/eclipse/orbit/bundles/com.google.guava/21.0.0-SNAPSHOT/com.google.guava-21.0.0-SNAPSHOT.jar can not be represented in Maven model and will not be visible to non-OSGi aware Maven plugins.
```

And indeed, a latter builds step fails because of this:

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.1:validate-classpath (default-validate-classpath) on project com.teamscale.ide: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.1:validate-classpath failed: bundleLocation not found: /home/sewe/.m2/repository/org/eclipse/orbit/bundles/com.google.guava/21.0.0-SNAPSHOT/com.google.guava-21.0.0-SNAPSHOT.jar -> [Help 1]

org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.1:validate-classpath (default-validate-classpath) on project com.teamscale.ide: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.1:validate-classpath failed: bundleLocation not found: /home/sewe/.m2/repository/org/eclipse/orbit/bundles/com.google.guava/21.0.0-SNAPSHOT/com.google.guava-21.0.0-SNAPSHOT.jar

Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.1:validate-classpath failed: bundleLocation not found: /home/sewe/.m2/repository/org/eclipse/orbit/bundles/com.google.guava/21.0.0-SNAPSHOT/com.google.guava-21.0.0-SNAPSHOT.jar

Caused by: java.lang.IllegalArgumentException: bundleLocation not found: /home/sewe/.m2/repository/org/eclipse/orbit/bundles/com.google.guava/21.0.0-SNAPSHOT/com.google.guava-21.0.0-SNAPSHOT.jar
```

What is really strange, however, and might hint at a bug in Tycho is that the very first build after a `rm -r ~/.m2/repository` succeeds, but executing the same `mvn` command again fails with the above warning and error. (Yes, not even the warning is shown during the initial run against the empty repository.)

AFAICT, this happens with Tycho 2.5.0, 2.6.0, 2.7.0, and 2.7.1.

Is there any advice on how to fix this or at least diagose this? I am not sure where the 21.0.0-SNAPSHOT version even comes from.

Best wishes,

Andreas Sewe

**Comment by laeubi** [comment]
This is probably due to bad maven-meta-data in an p2 update-site, can you provide a minimal project to reproduce the issue?

**Reply by sewe** [reply]
I pruned our production Eclipse plug-in (see https://github.com/sewe/tycho-discussion-877-reproducer) and have gotten quite far (no code, just `pom.xml`, `.target` and `MANIFEST.MF`). In particular, I have gotten to a point where the only p2 repository referenced is https://download.eclipse.org/releases/2021-12/202112081000. Not sure whether the bad metadata theory holds water then, as I expect the Simultaneus Release repository to be quite extensively used.

At any rate, here's how to reproduce the issue

```bash
git clone https://github.com/sewe/tycho-discussion-877-reproducer.git
cd tycho-discussion-877-reproducer

rm -r ~/.m2/repository # Clean slate

./mvnw clean verify # Succeeds
./mvnw clean verify # Fails
```

If you want to check with different Tycho versions, add `-DtychoVersion=2.7.1` or similar to the `mvnw` call. For me, the failures occur with Tycho 2.4.0 all the way up to 2.7.1. (Tycho 2.3.0 trips over a `Unknown OSGi execution environment: 'JavaSE-17'` error.)

**Reply by laeubi** [reply]
Alright, I'll take a look at it.

**Reply by sewe** [reply]
Thank you.

FWIW, an older target platform based on Eclipse 2021-09 doesn't exhibit the problem.

**Reply by sewe** [reply]
> FWIW, an older target platform based on Eclipse 2021-09 doesn't exhibit the problem.

Spoke to soon. Or rather, forgot that the problem occurs only on the **second** `mvnw` call. So, it fails for 2021-09, too.

**Reply by laeubi** [reply]
Running under linux I can't reproduce the issue with your example I can run the build as often as I wan't without seeing a failure. Do you have any special maven settings in use (e.g. p2 mirrors or something?).

**Reply by sewe** [reply]
No. My `~/.m2/settings.xml` is virtually empty. Also, I am under Linux as well.

```
$ ./mvnw --version
Apache Maven 3.8.3 (ff8e977a158738155dc465c6a97ffaf31982d739)
Maven home: /home/sewe/.m2/wrapper/dists/apache-maven-3.8.3-bin/5a6n1u8or3307vo2u2jgmkhm0t/apache-maven-3.8.3
Java version: 17.0.2, vendor: Debian, runtime: /usr/lib/jvm/java-17-openjdk-amd64
Default locale: en_CA, platform encoding: UTF-8
OS name: "linux", version: "5.10.0-13-amd64", arch: "amd64", family: "unix"
```

According to the log output and as expected, Maven dependencies all get downloaded from https://repo.maven.apache.org/maven2/, whereas p2 dependencies get downloaded from a variety of eclipse.org mirrors.

To ensure that I don't get bad metadata from a mirror, I've run everything (twice) with `-Dtycho.disableP2Mirrors` as well:
```
rm -r ~/.m2/repository # Clean slate

./mvnw clean verify -Dtycho.disableP2Mirrors=true # Succeeds
./mvnw clean verify -Dtycho.disableP2Mirrors=true # Fails
```
As expected, all p2 dependencies are now downloaded from https://download.eclipse.org/releases/2021-12/202112081000/. But the result is still the same: the second and any further builds fail.

However, downgrading the Java 11, the issue goes away:

```
$ ./mvnw --version
Apache Maven 3.8.3 (ff8e977a158738155dc465c6a97ffaf31982d739)
Maven home: /home/sewe/.m2/wrapper/dists/apache-maven-3.8.3-bin/5a6n1u8or3307vo2u2jgmkhm0t/apache-maven-3.8.3
Java version: 11.0.14, vendor: Debian, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_CA, platform encoding: UTF-8
OS name: "linux", version: "5.10.0-13-amd64", arch: "amd64", family: "unix"
```

@laeubi Can you try with Java 17 as well (assuming you were using a different version for your experiment)?

**Reply by laeubi** [reply]
That's really strange, it seems it even is enough to run the build once with java 11 to make it succeed. If I run it with Java 17 it fails on the second run. The only difference I can see it that for java > 14 we disable packed artifact support, you can try with Tycho 3.0.0-SNAPSHOT where we removed pack200 completely if that makes a difference.

**Reply by sewe** [reply]
The dropped pack200 support was indeed the hint I needed. Tried without `<includePackedArtifacts>` in the `target-platform-configuration` and it works even when all `mvnw` calls use Java 17.

I was assuming the option simply gets ignored on Java 17 (possibly emitting a warning), but this behaviour looks like a bug to me. Shall I create an issue from this discussion, @laeubi?

**Reply by laeubi** [reply]
I have created https://github.com/eclipse/tycho/issues/885 but not noticed the packSettings in the target configuration, this seems to be missing and an easy to fix.
